robottelo/report_portal/portal.py

```diff
@@ -0,0 +1,216 @@
+                For second latest, set latest = 2
+
+        :param int launch: The launch position in Report portal
+        """
+        super(Tests, self).__init__()
+        self.tests_url = f'{self.api_url}/item'
+        self.launch = self.launch(latest=launch)
+
+    @property
+    def defect_types(self):
+        """The various defect types in Report Report Portal for tests
+
+        This holds the codes for all the defect types, the tests will be marked with.
+
+        :returns dict: The dict of varios defect types and their codes
+        """
+        _defect_types = {
+            'product_bug': 'PB001',
+            'rhel_bug': 'RB001',
+            'satellite_bug': 'SB001',
+            'automation_bug': 'AB001',
+            'robottelo_bug': 'RB001',
+            'nailgun_bug': 'NG001',
+            'airgun_bug': 'AG001',
+            'system_issue': 'SI001',
+            'saucelabs_issue': 'SL001',
+            'to_investigate': 'TI001',
+            'no_defect': 'ND001',
+        }
+        return _defect_types
+
+    def _params(self, **kwargs):
+        """Parameters for Test items API request
+
+        :param dict kwargs: The tests `defect_type` and/or tests `status`
+
+        It contains filters to get the test items data from launch
+
+        :returns dict: parameters for API test items request
+        """
+        params = [
+            ('filter.eq.launch', self.launch['id']),
+            ('page.page', 1),
+            ('page.size', 50),
+            ('page.sort', 'start_time'),
+        ]
+        defect_type = kwargs.get('defect_type')
+        test_status = kwargs.get('status')
+        if defect_type in self.defect_types:
+            params.insert(0, ('filter.eq.issue$issue_type', self.defect_types[defect_type]))
+        if test_status:
+            params.insert(2, ('filter.eq.status', test_status))
+        return dict(params)
+
+    @retry(
+        stop=stop_after_attempt(4), wait=wait_fixed(10),
+    )
+    def _requester(self, params, page):
+        """The Test Items GET requester to fetch the data on a page
+
+        If any error and before Failing explicitly, it retries for 3 times with 10 seconds delay
+
+        :param dict params: The parameters dictionary for API request
+        :param int page: The Page number parameter manipulates the params
+
+        :returns tuple (int, list): Total pages count and the list of tests along with
+            each tests properties in a page
+        """
+        params['page.page'] = page
+        resp = requests.get(self.tests_url, headers=self.headers, params=params, verify=False)
+        resp.raise_for_status()
+        total_pages = resp.json()['page']['totalPages']
+        pagedata = resp.json().get('content')
+        return total_pages, pagedata
+
+    def _tests(self, params, data, page=1):
+        """The Recursive Function that roll over on all pages to get all tests data in an API request
```

Comment:
Is there a benefit here to this recursing?

It appears that this could be implemented with a non-recursing while statement.

```
total_pages, data = self._requester(params=params, page=page)
while page < total_pages:
    page += 1
    _, pagedata = self._requester(params=params, page=page)
    data.extend(pagedata)  # I'm not a fan of adding non-integers with +
return data
```

Reply:
I am not hard on using Recursion though I wanted to save some lines.

Reply:
Changing it to simple while loop.